src/Interpreters/InterpreterDropQuery.cpp

```diff
@@ -294,14 +294,24 @@ BlockIO InterpreterDropQuery::executeToTemporaryTable(const String & table_name,
 }
 
 
-BlockIO InterpreterDropQuery::executeToDatabase(const ASTDropQuery & query)
+BlockIO InterpreterDropQuery::executeToDatabase(ASTDropQuery & query)
 {
     DatabasePtr database;
     std::vector<UUID> tables_to_wait;
     BlockIO res;
     try
     {
-        res = executeToDatabaseImpl(query, database, tables_to_wait);
+        String db_name = getContext()->resolveTemporaryDatabase(query.getDatabase());
+        if (db_name == query.getDatabase())
+        {
+            res = executeToDatabaseImpl(query, database, tables_to_wait);
+        }
+        else
+        {
+            res = executeToTemporaryDatabase(query);
```

Comment:
Race condition is possible (executeToDatabaseImpl may be called for temporary database and vice versa)